### ISSUE TITLE
Counsel is incompatible with Emacs 24

### ISSUE THREAD

**Issue by basil-conto** [abo-abo/swiper#1579]
1. `make emacs=emacs24 plain`
2. <kbd>M-x</kbd>`counsel-linux-app`<kbd>RET</kbd>
   => `let: Symbol's function definition is void: directory-files-recursively`

This is due to `counsel-linux-apps-list-desktop-files` calling `directory-files-recursively`, which was only added in Emacs 25.1. Should `counsel-linux-apps-list-desktop-files` be reimplemented (e.g. using a compatibility shim), or should we wait until Emacs 26 is released and drop Emacs 24 support?

**Comment by basil-conto** [comment]
> I don't like the idea of dropping Emacs 24 support in the near future.

Fair enough; I must have wrongly gotten the impression from somewhere that you were considering raising the minimum supported version. Personally, I am in no hurry (wherever I can, I try to maintain compatibility with as far back as Emacs 23, as that is the oldest version installable from Debian's repositories).

> For many people using Debian Stretch or Ubuntu Xenial `emacs` is still the `emacs-24.3` binary.

Xenial ships 24.5, and Debian Stable (Stretch) has shipped both 25.1 and 24.5 for a while now. Neither of them includes 24.3.

> We can resolve the incompat in time.

Sure; hence the report.

> It would be nice for popular distros to start shipping either emacs-25 or emacs-26 in the base repo.

As mentioned, Debian Stable has shipped 25.1 for a while now. Artful is the first Ubuntu to ship Emacs 25, specifically 25.2.

**Comment by basil-conto** [comment]
Another incompatibility is `package-desc-extras`, which was added in Emacs 24.4.

**Comment by abo-abo** [comment]
> Another incompatibility is package-desc-extras, which was added in Emacs 24.4.

I think it's fine in this case, it's just one of the actions of a `counsel-` function.

**Comment by basil-conto** [comment]
> I also notice that `counsel-read-setq-expression` refers to `add-function`, `elisp-eldoc-documentation-function`, `elisp-completion-at-point`, etc. which were all added after Emacs 24.3.

See also #2708.
